File: kyulee/programmers-2/주차요금계산.java
import java.util.*;
import java.util.stream.Collectors;

public class 주차요금계산 {
    class Solution {
        public int[] solution(int[] fees, String[] records) {
            int baseTime = fees[0];
            int baseFee = fees[1];
            int unitTime = fees[2];
            int unitFee = fees[3];

            // 차량별 총 주차 시간 저장하는 맵
            Map<String, Integer> totalParkingTime = new HashMap<>();
            // 차량별 입차 시간 저장하는 맵
            Map<String, Integer> currentParkingTime = new HashMap<>();

            // 기록을 처리
            for (String record : records) {
                String[] parts = record.split(" ");
                int time = toMinutes(parts[0]);
                String carNumber = parts[1];
                boolean isEntry = parts[2].equals("IN");

                if (isEntry) {
                    currentParkingTime.put(carNumber, time);
                } else {
                    int entryTime = currentParkingTime.remove(carNumber);
                    int duration = time - entryTime;
                    totalParkingTime.put(carNumber, totalParkingTime.getOrDefault(carNumber, 0) + duration);
                }
            }

            // 23:59까지 출차하지 않은 차량 처리
            for (Map.Entry<String, Integer> entry : currentParkingTime.entrySet()) {
                int duration = toMinutes("23:59") - entry.getValue();
                totalParkingTime.put(entry.getKey(), totalParkingTime.getOrDefault(entry.getKey(), 0) + duration);
            }

            // 차량 번호별로 정렬
            List<String> carNumbers = new ArrayList<>(totalParkingTime.keySet());
            Collections.sort(carNumbers);

            // 결과를 저장할 배열
            int[] answer = new int[carNumbers.size()];
            int index = 0;

            // 주차 요금 계산
            for (String carNumber : carNumbers) {
                int parkingTime = totalParkingTime.get(carNumber);
                int fee = calculateFee(parkingTime, baseTime, baseFee, unitTime, unitFee);
                answer[index++] = fee;
            }

            return answer;
        }

        // 시간 문자열을 분 단위로 변환
        private int toMinutes(String time) {
            String[] parts = time.split(":");
            int hours = Integer.parseInt(parts[0]);
            int minutes = Integer.parseInt(parts[1]);
            return hours * 60 + minutes;
        }

        // 주차 요금 계산
        private int calculateFee(int parkingTime, int baseTime, int baseFee, int unitTime, int unitFee) {
            if (parkingTime <= baseTime) {
                return baseFee;
            }
            int extraTime = parkingTime - baseTime;
            int extraUnits = (int) Math.ceil((double) extraTime / unitTime);
            return baseFee + extraUnits * unitFee;
        }
    }
}

File: soyeong/programmers-2/주차-요금-계산.py
import math

def solution(fees, records):
    answer = []
    default_time, default_fee, unit_time, unit_fee = fees
    parking = {}
    using_time = {}

    # 기록 하나씩 처리
    for record in records:
        time, car_num, io = record.split()
        # 시간 문자열 > 숫자로 변환
        hour, minute = map(int, time.split(":"))
        time = hour * 60 + minute

        # 입차했으면
        if io == "IN":
            parking[car_num] = time
        elif io == "OUT":
            # 출차 시 후의 추가 시간이 있을때를 대비해 사용시간을 저장해둠
            if car_num in using_time:
                using_time[car_num] += (time - parking[car_num])
            else:
                using_time[car_num] = time - parking[car_num]
            # 현재 주차 된 기록은 삭제
            del parking[car_num]

    # 출차 기록이 없는 케이스 경우 IN 기록만 남아 있으므로 parking 기록을 조회해 파킹 제한시간에서 입차한 시간을 차감에 기록
    for car_num, time in parking.items():
        if car_num in using_time:
            using_time[car_num] += 1439 - time
        else:
            using_time[car_num] = 1439 - time

    # 차량번호 작은 순서대로 딕셔너리를 정렬 x[0] = car_num
    sorted_d = sorted(using_time.items(), key=lambda x: x[0])

    # 사용시간에 따른 요금 계산
    for car_num, time in sorted_d:
        # 기본요금 + 추가요금(올림((이용시간-기본시간)/단위시간))*요금)
        answer.append(default_fee + max(0, math.ceil((time - default_time) / unit_time)) * unit_fee)

    return answer